rust/cymbal/src/resolver.rs

```diff
@@ -8,35 +8,123 @@ use sourcemap::SourceMap;
 
 #[async_trait]
 pub trait Resolver: Send + Sync + 'static {
-    // Resolvers work on a per-frame basis, so we can be clever about the order
-    // in which we resolve them. We also force any resolver to handle all frame
-    // types
+    // TODO - I'm not totally convinced this resolver interface shouldn't enforce
+    // some kind of batch-style use (take a symbol set ref and a list of frame
+    // explicitly? I'm not sure)... right now this interface is prone to "holding it
+    // wrong" type performance issues. Resolvers should maybe even encode a "submit"
```

Comment:
Apple released an iphone a few years ago where if you held it a certain way the radio would stop working. Steve jobs [semi-famously defended the phone, saying users were "holding it wrong"](https://www.theguardian.com/technology/blog/2010/jun/25/iphone-reception-problems-solved). Of course, releasing a phone it's possible to hold so "wrong" it stops being a phone means, in reality, you've just built a phone wrong.

I use the phrase "holding it wrong" to refer to building something it's possible to use in a way that causes it to not work, or work so poorly it may as well not work - and as a signifier that it's not the user of the thing that's making the mistake there, but the designer of it (me). I don't know where I picked this habit up though.

Reply:
Ahh nice! I vaguely remember that - didn't they ultimately offer everyone a free case? Wonder what our equivalent is 🤔